File: services/api/src/main/java/com/launchly/project/services/RepositoryHintsService.java
package com.launchly.project.services;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.launchly.project.dto.RepositoryHintsResponse;
import com.launchly.project.entities.Project;
import org.springframework.stereotype.Service;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * Best-effort inference from public Git hosting raw files (no clone, no tokens).
 * Fills sensible defaults so users only need name + repo URL (+ deploy target when used).
 */
@Service
public class RepositoryHintsService {

    private static final Pattern GITHUB_HTTPS = Pattern.compile(
            "https?://github\\.com/([^/]+)/([^/]+?)(?:\\.git)?/?$", Pattern.CASE_INSENSITIVE);
    private static final Pattern GITHUB_SSH = Pattern.compile(
            "git@github\\.com:([^/]+)/([^/]+?)(?:\\.git)?", Pattern.CASE_INSENSITIVE);

    private static final Pattern GITLAB_HTTPS = Pattern.compile(
            "https?://gitlab\\.com/(.+?)(?:\\.git)?/?$", Pattern.CASE_INSENSITIVE);

    private static final Pattern GITEE_HTTPS = Pattern.compile(
            "https?://gitee\\.com/(.+?)(?:\\.git)?/?$", Pattern.CASE_INSENSITIVE);

    private static final Pattern PORT_IN_SCRIPT = Pattern.compile(
            "(?:--port|-p)\\s+(\\d{2,5})\\b|\\bPORT\\s*=\\s*(\\d{2,5})\\b");

    private static final Pattern README_INSTALL_LINE = Pattern.compile(
            "(?i)^\\s*((?:npm|pnpm|yarn)\\s+\\S+(?:\\s+\\S+)*)\\s*$");

    private final HttpClient http = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(8))
            .followRedirects(HttpClient.Redirect.NORMAL)
            .build();

    private final ObjectMapper mapper = new ObjectMapper();

    /**
     * Merge inferred values into {@code project} for any command/port/health field that is still blank.
     */
    public void fillBlanksFromRepository(Project project) {
        if (!hasText(project.getRepositoryUrl())) {
            return;
        }
        String branch = hasText(project.getDefaultBranch()) ? project.getDefaultBranch() : "main";
        RepositoryHintsResponse hints = infer(project.getRepositoryUrl().trim(), branch)
                .orElse(RepositoryHintsResponse.defaults());
        if (!hasText(project.getInstallCommand()) && hasText(hints.installCommand())) {
            project.setInstallCommand(hints.installCommand());
        }
        if (!hasText(project.getBuildCommand()) && hasText(hints.buildCommand())) {
            project.setBuildCommand(hints.buildCommand());
        }
        if (!hasText(project.getStartCommand()) && hasText(hints.startCommand())) {
            project.setStartCommand(hints.startCommand());
        }
        if (!hasText(project.getTestCommand()) && hasText(hints.testCommand())) {
            project.setTestCommand(hints.testCommand());
        }
        if (project.getDefaultPort() == null && hints.defaultPort() != null) {
            project.setDefaultPort(hints.defaultPort());
        }
        if (!hasText(project.getHealthCheckPath()) && hasText(hints.healthCheckPath())) {
            project.setHealthCheckPath(hints.healthCheckPath());
        }
    }

    public Optional<RepositoryHintsResponse> infer(String repositoryUrl, String branch) {
        if (!hasText(repositoryUrl) || !hasText(branch)) {
            return Optional.empty();
        }
        String refEnc = branch.replace("/", "%2F");
        Optional<String> pkgUrl = rawPackageJsonUrl(repositoryUrl.trim(), refEnc);
        if (pkgUrl.isEmpty()) {
            return Optional.empty();
        }
        Optional<String> jsonBody = httpGetString(pkgUrl.get());
        if (jsonBody.isEmpty()) {
            return Optional.empty();
        }
        try {
            JsonNode root = mapper.readTree(jsonBody.get());
            PackageManager pm = detectPackageManager(root, repositoryUrl.trim(), refEnc);

            JsonNode scripts = root.path("scripts");
            String startScript = textOrNull(scripts, "start");
            String buildScript = textOrNull(scripts, "build");
            String testScript = textOrNull(scripts, "test");

            String install = installCommand(pm);
            String build = buildScript != null ? runScript(pm, "build") : null;
            String start = startCommand(pm);
            String test = testScript != null ? runScript(pm, "test") : null;

            Integer port = parsePortFromStart(startScript);
            if (port == null) {
                port = 3000;
            }

            String source = "package.json";
            Optional<String> readmeLine = readmeInstallLine(repositoryUrl.trim(), refEnc);
            if (readmeLine.isPresent()
                    && pm == PackageManager.NPM
                    && (readmeLine.get().contains("pnpm") || readmeLine.get().contains("yarn"))) {
                install = readmeLine.get();
                source = "package.json+readme";
            }

            return Optional.of(new RepositoryHintsResponse(
                    install,
                    build,
                    start,
                    test,
                    port,
                    null,
                    source
            ));
        } catch (Exception e) {
            return Optional.empty();
        }
    }

    private Optional<String> readmeInstallLine(String repositoryUrl, String refEnc) {
        Optional<String> readmeUrl = rawReadmeUrl(repositoryUrl, refEnc);
        if (readmeUrl.isEmpty()) {
            return Optional.empty();
        }
        Optional<String> body = httpGetString(readmeUrl.get());
        if (body.isEmpty()) {
            return Optional.empty();
        }
        for (String line : body.get().split("\n")) {
            Matcher m = README_INSTALL_LINE.matcher(line);
            if (m.matches()) {
                return Optional.of(m.group(1).trim());
            }
        }
        return Optional.empty();
    }

    private Optional<String> rawReadmeUrl(String repositoryUrl, String refEnc) {
        Matcher gh = GITHUB_HTTPS.matcher(repositoryUrl);
        if (gh.matches()) {
            return Optional.of(String.format(
                    "https://raw.githubusercontent.com/%s/%s/%s/README.md",
                    gh.group(1), gh.group(2), refEnc));
        }
        Matcher ghs = GITHUB_SSH.matcher(repositoryUrl);
        if (ghs.matches()) {
            return Optional.of(String.format(
                    "https://raw.githubusercontent.com/%s/%s/%s/README.md",
                    ghs.group(1), ghs.group(2), refEnc));
        }
        Matcher gl = GITLAB_HTTPS.matcher(repositoryUrl);
        if (gl.matches()) {
            return Optional.of(String.format(
                    "https://gitlab.com/%s/-/raw/%s/README.md",
                    gl.group(1), refEnc));
        }
        Matcher ge = GITEE_HTTPS.matcher(repositoryUrl);
        if (ge.matches()) {
            return Optional.of(String.format(
                    "https://gitee.com/%s/raw/%s/README.md",
                    ge.group(1), refEnc));
        }
        return Optional.empty();
    }

    private Optional<String> rawPackageJsonUrl(String repositoryUrl, String refEnc) {
        Matcher gh = GITHUB_HTTPS.matcher(repositoryUrl);
        if (gh.matches()) {
            return Optional.of(String.format(
                    "https://raw.githubusercontent.com/%s/%s/%s/package.json",
                    gh.group(1), gh.group(2), refEnc));
        }
        Matcher ghs = GITHUB_SSH.matcher(repositoryUrl);
        if (ghs.matches()) {
            return Optional.of(String.format(
                    "https://raw.githubusercontent.com/%s/%s/%s/package.json",
                    ghs.group(1), ghs.group(2), refEnc));
        }
        Matcher gl = GITLAB_HTTPS.matcher(repositoryUrl);
        if (gl.matches()) {
            return Optional.of(String.format(
                    "https://gitlab.com/%s/-/raw/%s/package.json",
                    gl.group(1), refEnc));
        }
        Matcher ge = GITEE_HTTPS.matcher(repositoryUrl);
        if (ge.matches()) {
            return Optional.of(String.format(
                    "https://gitee.com/%s/raw/%s/package.json",
                    ge.group(1), refEnc));
        }
        return Optional.empty();
    }

    private enum PackageManager {
        NPM, PNPM, YARN
    }

    private PackageManager detectPackageManager(JsonNode root, String repositoryUrl, String refEnc) {
        String pmField = textOrNull(root, "packageManager");
        if (pmField != null) {
            if (pmField.startsWith("pnpm")) {
                return PackageManager.PNPM;
            }
            if (pmField.startsWith("yarn")) {
                return PackageManager.YARN;
            }
        }
        Optional<String> base = rawPackageJsonUrl(repositoryUrl, refEnc);
        if (base.isPresent()) {
            String u = base.get();
            int slash = u.lastIndexOf('/');
            String prefix = u.substring(0, slash + 1);
            if (resourceExists(prefix + "pnpm-lock.yaml")) {
                return PackageManager.PNPM;
            }
            if (resourceExists(prefix + "yarn.lock")) {
                return PackageManager.YARN;
            }
        }
        return PackageManager.NPM;
    }

    private boolean resourceExists(String url) {
        try {
            HttpRequest req = HttpRequest.newBuilder(URI.create(url))
                    .method("HEAD", HttpRequest.BodyPublishers.noBody())
                    .timeout(Duration.ofSeconds(8))
                    .header("User-Agent", "Launchly-RepositoryHints/1.0")
                    .build();
            HttpResponse<Void> res = http.send(req, HttpResponse.BodyHandlers.discarding());
            if (res.statusCode() == 200) {
                return true;
            }
            if (res.statusCode() == 405 || res.statusCode() == 404) {
                return httpGetString(url).map(s -> !s.isBlank()).orElse(false);
            }
        } catch (Exception ignored) {
            return httpGetString(url).map(s -> !s.isBlank()).orElse(false);
        }
        return false;
    }

    private String installCommand(PackageManager pm) {
        return switch (pm) {
            case PNPM -> "corepack enable && pnpm install --frozen-lockfile";
            case YARN -> "corepack enable && yarn install --immutable";
            case NPM -> "npm ci --omit=dev || npm install --omit=dev";
        };
    }

    private String runScript(PackageManager pm, String script) {
        return switch (pm) {
            case PNPM -> "pnpm run " + script;
            case YARN -> "yarn " + script;
            case NPM -> "npm run " + script;
        };
    }

    private String startCommand(PackageManager pm) {
        return switch (pm) {
            case PNPM -> "pnpm start";
            case YARN -> "yarn start";
            case NPM -> "npm start";
        };
    }

    private Integer parsePortFromStart(String startScript) {
        if (startScript == null) {
            return null;
        }
        Matcher m = PORT_IN_SCRIPT.matcher(startScript);
        if (m.find()) {
            String g = m.group(1) != null ? m.group(1) : m.group(2);
            try {
                return Integer.parseInt(g);
            } catch (NumberFormatException ignored) {
                return null;
            }
        }
        return null;
    }

    private String textOrNull(JsonNode parent, String field) {
        JsonNode n = parent.get(field);
        if (n == null || n.isNull() || !n.isTextual()) {
            return null;
        }
        String t = n.asText().trim();
        return t.isEmpty() ? null : t;
    }

    private Optional<String> httpGetString(String url) {
        try {
            HttpRequest req = HttpRequest.newBuilder(URI.create(url))
                    .timeout(Duration.ofSeconds(15))
                    .header("User-Agent", "Launchly-RepositoryHints/1.0")
                    .GET()
                    .build();
            HttpResponse<String> res = http.send(req, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8));
            if (res.statusCode() >= 200 && res.statusCode() < 300) {
                String body = res.body();
                if (body != null && body.length() > 2_000_000) {
                    return Optional.empty();
                }
                return Optional.ofNullable(body);
            }
        } catch (Exception ignored) {
            return Optional.empty();
        }
        return Optional.empty();
    }

    private static boolean hasText(String s) {
        return s != null && !s.isBlank();
    }
}

File: services/api/src/project/repository-hints.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { PrismaService } from '../common/prisma/prisma.service';
import { RepositoryHintsResponse } from './dto';

type PackageManager = 'NPM' | 'PNPM' | 'YARN';

const GITHUB_HTTPS = /https?:\/\/github\.com\/([^/]+)\/([^/]+?)(?:\.git)?\/?$/i;
const GITHUB_SSH = /git@github\.com:([^/]+)\/([^/]+?)(?:\.git)?/i;
const GITLAB_HTTPS = /https?:\/\/gitlab\.com\/(.+?)(?:\.git)?\/?$/i;
const GITEE_HTTPS = /https?:\/\/gitee\.com\/(.+?)(?:\.git)?\/?$/i;
const PORT_IN_SCRIPT = /(?:--port|-p)\s+(\d{2,5})\b|\bPORT\s*=\s*(\d{2,5})\b/;
const README_INSTALL_LINE = /^\s*((?:npm|pnpm|yarn)\s+\S+(?:\s+\S+)*)\s*$/;

@Injectable()
export class RepositoryHintsService {
  private readonly logger = new Logger(RepositoryHintsService.name);

  constructor(private readonly prisma: PrismaService) {}

  /**
   * Merge inferred values into project fields that are still blank.
   */
  async fillBlanksFromRepository(project: {
    repositoryUrl?: string | null;
    defaultBranch?: string;
    installCommand?: string | null;
    buildCommand?: string | null;
    startCommand?: string | null;
    testCommand?: string | null;
    defaultPort?: number | null;
    healthCheckPath?: string | null;
  }): Promise<void> {
    if (!project.repositoryUrl) return;
    const branch = project.defaultBranch || 'main';
    const hints = await this.infer(project.repositoryUrl.trim(), branch);
    if (!hints) return;
    if (!project.installCommand && hints.installCommand) project.installCommand = hints.installCommand;
    if (!project.buildCommand && hints.buildCommand) project.buildCommand = hints.buildCommand;
    if (!project.startCommand && hints.startCommand) project.startCommand = hints.startCommand;
    if (!project.testCommand && hints.testCommand) project.testCommand = hints.testCommand;
    if (project.defaultPort == null && hints.defaultPort != null) project.defaultPort = hints.defaultPort;
    if (!project.healthCheckPath && hints.healthCheckPath) project.healthCheckPath = hints.healthCheckPath;
  }

  async infer(repositoryUrl: string, branch: string): Promise<RepositoryHintsResponse | null> {
    if (!repositoryUrl || !branch) return null;
    const refEnc = branch.replace(/\//g, '%2F');
    const pkgUrl = this.rawPackageJsonUrl(repositoryUrl, refEnc);
    if (!pkgUrl) return null;

    const jsonBody = await this.httpGetString(pkgUrl);
    if (!jsonBody) return null;

    try {
      const root = JSON.parse(jsonBody);
      const pm = await this.detectPackageManager(root, repositoryUrl, refEnc);
      const scripts = root.scripts || {};
      const startScript: string | undefined = scripts.start;
      const buildScript: string | undefined = scripts.build;
      const testScript: string | undefined = scripts.test;

      let install = this.installCommand(pm);
      const build = buildScript ? this.runScript(pm, 'build') : null;
      const start = this.startCommand(pm);
      const test = testScript ? this.runScript(pm, 'test') : null;
      let port = this.parsePortFromStart(startScript) ?? 3000;

      let source = 'package.json';
      const readmeLine = await this.readmeInstallLine(repositoryUrl, refEnc);
      if (readmeLine && pm === 'NPM' && (readmeLine.includes('pnpm') || readmeLine.includes('yarn'))) {
        install = readmeLine;
        source = 'package.json+readme';
      }

      return { installCommand: install, buildCommand: build, startCommand: start, testCommand: test, defaultPort: port, healthCheckPath: null, source };
    } catch {
      return null;
    }
  }

  private async readmeInstallLine(repositoryUrl: string, refEnc: string): Promise<string | null> {
    const readmeUrl = this.rawReadmeUrl(repositoryUrl, refEnc);
    if (!readmeUrl) return null;
    const body = await this.httpGetString(readmeUrl);
    if (!body) return null;
    for (const line of body.split('\n')) {
      const m = line.match(README_INSTALL_LINE);
      if (m) return m[1].trim();
    }
    return null;
  }

  private rawReadmeUrl(repositoryUrl: string, refEnc: string): string | null {
    let m = GITHUB_HTTPS.exec(repositoryUrl);
    if (m) return `https://raw.githubusercontent.com/${m[1]}/${m[2]}/${refEnc}/README.md`;
    m = GITHUB_SSH.exec(repositoryUrl);
    if (m) return `https://raw.githubusercontent.com/${m[1]}/${m[2]}/${refEnc}/README.md`;
    m = GITLAB_HTTPS.exec(repositoryUrl);
    if (m) return `https://gitlab.com/${m[1]}/-/raw/${refEnc}/README.md`;
    m = GITEE_HTTPS.exec(repositoryUrl);
    if (m) return `https://gitee.com/${m[1]}/raw/${refEnc}/README.md`;
    return null;
  }

  private rawPackageJsonUrl(repositoryUrl: string, refEnc: string): string | null {
    let m = GITHUB_HTTPS.exec(repositoryUrl);
    if (m) return `https://raw.githubusercontent.com/${m[1]}/${m[2]}/${refEnc}/package.json`;
    m = GITHUB_SSH.exec(repositoryUrl);
    if (m) return `https://raw.githubusercontent.com/${m[1]}/${m[2]}/${refEnc}/package.json`;
    m = GITLAB_HTTPS.exec(repositoryUrl);
    if (m) return `https://gitlab.com/${m[1]}/-/raw/${refEnc}/package.json`;
    m = GITEE_HTTPS.exec(repositoryUrl);
    if (m) return `https://gitee.com/${m[1]}/raw/${refEnc}/package.json`;
    return null;
  }

  private async detectPackageManager(root: any, repositoryUrl: string, refEnc: string): Promise<PackageManager> {
    const pmField: string | undefined = root.packageManager;
    if (pmField) {
      if (pmField.startsWith('pnpm')) return 'PNPM';
      if (pmField.startsWith('yarn')) return 'YARN';
    }
    const base = this.rawPackageJsonUrl(repositoryUrl, refEnc);
    if (base) {
      const slash = base.lastIndexOf('/');
      const prefix = base.substring(0, slash + 1);
      if (await this.resourceExists(prefix + 'pnpm-lock.yaml')) return 'PNPM';
      if (await this.resourceExists(prefix + 'yarn.lock')) return 'YARN';
    }
    return 'NPM';
  }

  private async resourceExists(url: string): Promise<boolean> {
    try {
      const res = await fetch(url, { method: 'HEAD', signal: AbortSignal.timeout(8000) });
      if (res.ok) return true;
      if (res.status === 405 || res.status === 404) {
        const body = await this.httpGetString(url);
        return !!body && body.trim().length > 0;
      }
    } catch {
      const body = await this.httpGetString(url);
      return !!body && body.trim().length > 0;
    }
    return false;
  }

  private installCommand(pm: PackageManager): string {
    switch (pm) {
      case 'PNPM': return 'corepack enable && pnpm install --frozen-lockfile';
      case 'YARN': return 'corepack enable && yarn install --immutable';
      case 'NPM': return 'npm ci --omit=dev || npm install --omit=dev';
    }
  }

  private runScript(pm: PackageManager, script: string): string {
    switch (pm) {
      case 'PNPM': return `pnpm run ${script}`;
      case 'YARN': return `yarn ${script}`;
      case 'NPM': return `npm run ${script}`;
    }
  }

  private startCommand(pm: PackageManager): string {
    switch (pm) {
      case 'PNPM': return 'pnpm start';
      case 'YARN': return 'yarn start';
      case 'NPM': return 'npm start';
    }
  }

  private parsePortFromStart(startScript?: string): number | null {
    if (!startScript) return null;
    const m = startScript.match(PORT_IN_SCRIPT);
    if (m) {
      const g = m[1] ?? m[2];
      const n = parseInt(g, 10);
      return isNaN(n) ? null : n;
    }
    return null;
  }

  private async httpGetString(url: string): Promise<string | null> {
    try {
      const res = await fetch(url, {
        signal: AbortSignal.timeout(15000),
        headers: { 'User-Agent': 'Launchly-RepositoryHints/1.0' },
      });
      if (res.ok) {
        const body = await res.text();
        if (body.length > 2_000_000) return null;
        return body;
      }
    } catch {
      return null;
    }
    return null;
  }
}
